lxd/acme/acme.go

```diff
@@ -0,0 +1,179 @@
+package acme
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/go-acme/lego/acme"
+	"github.com/go-acme/lego/certcrypto"
+	"github.com/go-acme/lego/certificate"
+	"github.com/go-acme/lego/lego"
+	"github.com/go-acme/lego/registration"
+
+	"github.com/lxc/lxd/lxd/state"
+	"github.com/lxc/lxd/lxd/util"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/logger"
+)
+
+// retries describes the number of retries after which LXD will give up registering a user or
+// issuing a new certificate. The number 5 was chosen because Let's Encrypt has a limit of 5
+// failures per account, per hostname, per hour.
+const retries = 5
+
+// UpdateCertificate updates the certificate.
+func UpdateCertificate(s *state.State, provider HTTP01Provider, clustered bool, domain string, email string, caURL string) (*certificate.Resource, error) {
+	clusterCertFilename := shared.VarPath("cluster.crt.new")
+
+	// If cluster.crt.new exists, it means that a previously issued certificate couldn't be
+	// distributed to all cluster members and was therefore kept back. In this case, don't issue
+	// a new certificate but return the previously issued one.
+	if clustered && shared.PathExists(clusterCertFilename) {
+		keyFilename := shared.VarPath("cluster.key")
+
+		clusterCert, err := os.ReadFile(clusterCertFilename)
+		if err != nil {
+			return nil, fmt.Errorf("Failed reading cluster certificate file: %w", err)
+		}
+
+		key, err := os.ReadFile(keyFilename)
+		if err != nil {
+			return nil, fmt.Errorf("Failed reading cluster key file: %w", err)
+		}
+
+		keyPair, err := tls.X509KeyPair(clusterCert, key)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to get keypair: %w", err)
+		}
+
+		cert, err := x509.ParseCertificate(keyPair.Certificate[0])
+		if err != nil {
+			return nil, fmt.Errorf("Failed to parse certificate: %w", err)
+		}
+		// If the domain matches the certificate's DNS names, and the certificate is valid for more
+		// than 30 days, return it.
+		if shared.StringInSlice(domain, cert.DNSNames) && !time.Now().After(cert.NotAfter.Add(-time.Hour*24*30)) {
```

Comment:
As we have this time logic in 2 places now, it feels like we should have the time variable stored as a constant?

Reply:
How about we move this to a separate function `certificateNeedsUpdate` as the entire condition is in 2 places.

Reply:
Sounds good to me :)